analytics_dashboard/core/middleware.py

```diff
@@ -8,10 +8,24 @@
 from django.template.response import TemplateResponse
 
 from core.exceptions import ServiceUnavailableError
+from waffle import flag_is_active
 
 logger = logging.getLogger(__name__)
 
 
+# For https://tasks.opencraft.com/browse/SE-3950 . If an upstream solution is invented by the time you're rebasing this,
```

Comment:
@Kelketek even though the solution looks good, I can't accept it because it would be introducing additional code drift to _edx-olive_. 

We'll need an upstream pull request to be created in order to upstream the changes to `master`. And the pull request to _edx-olive_ should be a backport for that change.

Reply:
We cannot upstream this code in this fashion because edx-insights upstream uses a more recent version of Django, which changes how middleware works. Also upstream uses the Django plugin utils. The purpose of this PR is to get the changes available in a way that Campus can use today. We will then make a follow-up PR to add the changes to upstream in a way that's compatible with the newer versions of Django and which passes edX's requirements.

We do not, at least last I checked, wait for the upstream version to be merged before getting the client's instance patched. Instead, we fix the client version, then upstream the changes, which can take several weeks. Once the changes are upstreamed, we either make a commit to remove the changes we made and update to the latest upstream version, or we wait until the next rebase and remove the commit as part of the process.

The comment above makes it easy for a person doing such a rebase to verify if we got upstream to merge the change in time.

Reply:
> The purpose of this PR is to get the changes available in a way that Campus can use today. We will then make a follow-up PR to add the changes to upstream in a way that's compatible with the newer versions of Django and which passes edX's requirements.

That sounds reasonable, as long as it is done in the scope of the same ticket.
As long as there are changes which can be backported to a different release, we can create such a Pull Request for `edx-olive` for now.

But it should be indicated in the pull requested that this change should be dropped once we upgrade them to a new release. And we'll need to indicate which pull request we'll need to backport when upgrading them to a new release. 

> We do not, at least last I checked, wait for the upstream version to be merged before getting the client's instance patched. Instead, we fix the client version, then upstream the changes, which can take several weeks. Once the changes are upstreamed, we either make a commit to remove the changes we made and update to the latest upstream version, or we wait until the next rebase and remove the commit as part of the process.

You're absolutely right, I'm sorry for not explaining myself properly. What I was referring to is preparing the changes for `master` and then backporting them to the specific release, before opening the pull request. 

Then the changes get reviewed internally on the backported pull request, once approved, any changes requested there get applied to the upstream pull request, and then the upstream pull request gets reviewed by an internal member. 

But what I was referring to mainly is backporting the changes that were applied to `master`.

However, if we do as you mentioned, and apply something temporary which we'd throw away later on, then we can consider making two separate changes. However, I'd like the opinions of @itsjeyd and/or @clemente regarding this.

Reply:
@Kelketek @nizarmah 

> We do not, at least last I checked, wait for the upstream version to be merged before getting the client's instance patched. 

This depends on each client. With some we do {upstream fix + client backport} at the same time. With Campus, from what I know from @itsjeyd, we *do* wait for the fix to be merged upstream. Some time ago I added this particular information to the epic description (SE-311), please check the `Upstreaming strategy` section there.
However, for urgent changes we may do the extra work of backporting. I don't know who decides when to do it (CC @itsjeyd).
I like always backporting; otherwise the client must wait for months. But we may need to follow the standard Campus approach. We can keep the general upstreaming discussion at SE-311.

Reply:
Thanks for the clarity here, @clemente . I actually need to close this PR, though, since as discussed in the ticket we didn't end up needing it anyway. :/ We can continue discussion about the upstream PR there, though, since it still has some useful attributes.